File: backend/services/bdi_calculator.py
"""
Cálculo BDI conforme fórmula TCU (Acórdão 2622/2013).
BDI = [(1 + AC + S + R + DF) × (1 + L) / (1 - T) - 1] × 100
"""

from __future__ import annotations

from typing import Any, Dict, List


CATEGORIA_LABELS = {
    "despesas_indiretas": "Despesas indiretas",
    "risco": "Risco",
    "lucro": "Lucro",
    "tributos": "Tributos",
}


def _find_valor(componentes: List[Dict[str, Any]], ids: List[str]) -> float:
    for comp_id in ids:
        for comp in componentes:
            if str(comp.get("id", "")) == comp_id:
                return float(comp.get("valor") or 0) / 100.0
    return 0.0


def calcular_bdi_tcu(componentes: List[Dict[str, Any]]) -> Dict[str, Any]:
    ac = _find_valor(componentes, ["administracao_central"])
    s = _find_valor(componentes, ["seguro_garantia"])
    r = _find_valor(componentes, ["risco"])
    df = _find_valor(componentes, ["despesas_financeiras"])
    lucro = _find_valor(componentes, ["lucro"])

    tributos = sum(
        float(c.get("valor") or 0) / 100.0
        for c in componentes
        if str(c.get("categoria", "")) == "tributos"
    )

    numerator = (1 + ac + s + r + df) * (1 + lucro)
    denominator = 1 - tributos
    fator = numerator / denominator if denominator > 0 else numerator
    bdi_percentual = round((fator - 1) * 100, 2)
    fator_bdi = round(1 + bdi_percentual / 100, 4)

    breakdown_map: Dict[str, float] = {}
    for comp in componentes:
        cat = str(comp.get("categoria") or "")
        label = CATEGORIA_LABELS.get(cat, cat)
        breakdown_map[label] = breakdown_map.get(label, 0.0) + float(comp.get("valor") or 0)

    breakdown = [
        {"categoria": cat, "total": round(total, 2)}
        for cat, total in breakdown_map.items()
    ]

    return {
        "bdi_percentual": bdi_percentual,
        "fator_bdi": fator_bdi,
        "breakdown": breakdown,
    }

File: frontend/src/utils/bdiCalculator.ts
import type { BDIComponente, BDIResultado } from "../types/bdi";

const CATEGORIA_LABELS: Record<string, string> = {
  despesas_indiretas: "Despesas indiretas",
  risco: "Risco",
  lucro: "Lucro",
  tributos: "Tributos",
};

function findValor(componentes: BDIComponente[], ids: string[]): number {
  for (const id of ids) {
    const found = componentes.find((c) => c.id === id);
    if (found) return found.valor / 100;
  }
  return 0;
}

/** Fórmula TCU: [(1+AC+S+R+DF)×(1+L)/(1-T)-1]×100 */
export function calcularBDI(componentes: BDIComponente[]): BDIResultado {
  const AC = findValor(componentes, ["administracao_central"]);
  const S = findValor(componentes, ["seguro_garantia"]);
  const R = findValor(componentes, ["risco"]);
  const DF = findValor(componentes, ["despesas_financeiras"]);
  const L = findValor(componentes, ["lucro"]);

  const T = componentes
    .filter((c) => c.categoria === "tributos")
    .reduce((sum, c) => sum + c.valor / 100, 0);

  const numerator = (1 + AC + S + R + DF) * (1 + L);
  const denominator = 1 - T;
  const fator = denominator > 0 ? numerator / denominator : numerator;
  const bdiPercentual = Math.round((fator - 1) * 10000) / 100;
  const fatorBDI = Math.round((1 + bdiPercentual / 100) * 10000) / 10000;

  const breakdownMap: Record<string, number> = {};
  for (const comp of componentes) {
    const label = CATEGORIA_LABELS[comp.categoria] ?? comp.categoria;
    breakdownMap[label] = (breakdownMap[label] ?? 0) + comp.valor;
  }

  const breakdown = Object.entries(breakdownMap).map(([categoria, total]) => ({
    categoria,
    total: Math.round(total * 100) / 100,
  }));

  return { bdiPercentual, fatorBDI, breakdown };
}

export function formatBDIPercent(value: number): string {
  return `${value.toLocaleString("pt-BR", {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  })}%`;
}

export function formatCurrencyBRL(value: number): string {
  return value.toLocaleString("pt-BR", { style: "currency", currency: "BRL" });
}
